packages/frontend-plugin-api/src/analytics/Tracker.ts

```diff
@@ -0,0 +1,180 @@
+    value?: number;
+    attributes?: AnalyticsEventAttributes;
+    context: AnalyticsContextValue;
+  };
+  /**
+   * Stores the most recent routable extension render.
+   */
+  mostRecentRoutableExtensionRender?: {
+    context: AnalyticsContextValue;
+  };
+  /**
+   * Tracks whether or not a beforeunload event listener has already been
+   * registered.
+   */
+  beforeUnloadRegistered: boolean;
+};
+
+/**
+ * Temporary global store for select event data. Used to make `navigate` events
+ * more accurate when gathered mountpoints are used.
+ */
+const globalEvents = getOrCreateGlobalSingleton<TempGlobalEvents>(
+  'core-plugin-api:analytics-tracker-events',
+  () => ({
+    mostRecentGatheredNavigation: undefined,
+    mostRecentRoutableExtensionRender: undefined,
+    beforeUnloadRegistered: false,
+  }),
+);
+
+/**
+ * Internal-only event representing when a routable extension is rendered.
+ */
+export const routableExtensionRenderedEvent = '_ROUTABLE-EXTENSION-RENDERED';
+
+export class Tracker implements AnalyticsTracker {
+  constructor(
+    private readonly analyticsApi: AnalyticsApi,
+    private context: AnalyticsContextValue = {
+      routeRef: 'unknown',
+      pluginId: 'root',
+      extension: 'App',
+    },
+  ) {
+    // Only register a single beforeunload event across all trackers.
+    if (!globalEvents.beforeUnloadRegistered) {
+      // Before the page unloads, attempt to capture any deferred navigation
+      // events that haven't yet been captured.
+      addEventListener(
+        'beforeunload',
+        () => {
+          if (globalEvents.mostRecentGatheredNavigation) {
+            this.analyticsApi.captureEvent({
+              ...globalEvents.mostRecentGatheredNavigation,
+              ...globalEvents.mostRecentRoutableExtensionRender,
+            });
+            globalEvents.mostRecentGatheredNavigation = undefined;
+            globalEvents.mostRecentRoutableExtensionRender = undefined;
+          }
+        },
+        { once: true, passive: true },
+      );
+
+      // Prevent duplicate handlers from being registered.
+      globalEvents.beforeUnloadRegistered = true;
+    }
+  }
+
+  setContext(context: AnalyticsContextValue) {
+    this.context = context;
+  }
+
+  captureEvent(
+    action: string,
+    subject: string,
+    {
+      value,
+      attributes,
+    }: { value?: number; attributes?: AnalyticsEventAttributes } = {},
+  ) {
+    // Never pass internal "_routeNodeType" context value.
+    const { _routeNodeType, ...context } = this.context;
+
+    // Never fire the special "_routable-extension-rendered" internal event.
+    if (action === routableExtensionRenderedEvent) {
+      // But keep track of it if we're delaying a `navigate` event for a
+      // a gathered route node type.
+      if (globalEvents.mostRecentGatheredNavigation) {
+        globalEvents.mostRecentRoutableExtensionRender = {
+          context: {
+            ...context,
+            extension: 'App',
```

Comment:
Looking through this I very much get the feeling that we need to bring extensions into the mix here straight away, or there's a significant risk that some of the logic is lost here.

Reply:
Updated https://github.com/backstage/backstage/pull/21244/commits/5c794cd5835c6c2be565134d0a744743dc95637a